src/sentry/migrations/0869_backfill_open_periods.py

```diff
@@ -0,0 +1,205 @@
+
+    for activity in activities:
+        if activity.type == ActivityType.SET_RESOLVED.value:
+            end = activity.datetime
+            end_activity = activity
+        elif activity.type == ActivityType.SET_REGRESSION.value:
+            start = activity.datetime
+            if end is None:
+                logger.error(
+                    "No end activity found for group open period backfill",
+                    extra={"group_id": group_id, "starting_activity": activity.id},
+                )
+            if start is not None and end is not None:
+                open_periods.append(
+                    GroupOpenPeriod(
+                        group_id=group_id,
+                        project_id=project_id,
+                        date_started=start,
+                        date_ended=end,
+                        resolution_activity=end_activity,
+                        user_id=end_activity.user_id if end_activity else None,
+                    )
+                )
+                end = None
+                end_activity = None
+    # Add the very first open period, which has no UNRESOLVED activity for the group creation
+    open_periods.append(
+        GroupOpenPeriod(
+            group_id=group_id,
+            project_id=project_id,
+            date_started=first_seen,
+            date_ended=end if end else None,
+            resolution_activity=end_activity,
+            user_id=end_activity.user_id if end_activity else None,
+        )
+    )
+    return open_periods
+
+
+def _backfill_group_open_periods(
+    apps: StateApps, group_data: list[tuple[int, datetime, int, int]]
+) -> None:
+    GroupOpenPeriod = apps.get_model("sentry", "GroupOpenPeriod")
+    Activity = apps.get_model("sentry", "Activity")
+
+    group_ids = [group_id for group_id, _, _, _ in group_data]
+    groups_with_open_periods = set(
+        GroupOpenPeriod.objects.filter(group_id__in=group_ids).values_list("group_id", flat=True)
+    )
+
+    # Filter the relevant activities for groups that need to be backfilled.
+    activities = defaultdict(list)
+    for activity in Activity.objects.filter(
+        group_id__in=group_ids,
+        type__in=[ActivityType.SET_REGRESSION.value, ActivityType.SET_RESOLVED.value],
+    ).order_by("-datetime"):
+        if activity.group_id not in groups_with_open_periods:
+            activities[activity.group_id].append(activity)
+
+    open_periods = []
+    for group_id, first_seen, status, project_id in group_data:
+        # Skip groups that already have open periods
+        if group_id in groups_with_open_periods:
+            continue
+
+        open_periods.extend(
+            get_open_periods_for_group(
+                apps,
+                group_id,
+                status,
+                project_id,
+                first_seen,
+                activities[group_id],
+                GroupOpenPeriod,
+            )
+        )
+
+    GroupOpenPeriod.objects.bulk_create(open_periods)
```

Comment:
It might be a good idea to wrap this in a transaction so that we don't end up in a partial state if something fails.